### PR TITLE
The Old Prince: 1871

### PR DESCRIPTION
**Pull request by kelsin**
Prealpha state. Games can be played until completion.

Currently the open bugs I know about are here: https://github.com/kelsin/18xx/issues

### CODE REVIEW

File: lib/engine/game/g_1871/meta.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+        DEV_STAGE = :prealpha
+
+        GAME_TITLE = 'The Old Prince'.freeze
+        GAME_SUBTITLE = '1871'.freeze
+        GAME_FULL_TITLE = 'The Old Prince: 1871'.freeze
+        GAME_DROPDOWN_TITLE = 'The Old Prince: 1871'.freeze
+
+        GAME_DESIGNER = 'Lucas Boyd'.freeze
+        GAME_IMPLEMENTER = 'Christopher Giroir'.freeze
```
Comment (roseundy) [inline lib/engine/game/g_1871/meta.rb:19]: What is this used for?

Comment (kelsin) [reply lib/engine/game/g_1871/meta.rb:19]: I filled it out since it was in the base meta file: https://github.com/tobymao/18xx/blob/master/lib/engine/game/meta.rb#L21

Looks like it's displayed in with the game info:
https://github.com/tobymao/18xx/blob/master/assets/app/view/game/game_meta.rb#L14

Comment (crericha) [reply lib/engine/game/g_1871/meta.rb:19]: I added it for 1828 because the author wanted is to use the licensing model instead of providing permission. The 18xx.games version is not associated with JC and this was the way Toby and I came up with to associate it with someone (me).

I'm not sure if any other games use it.

Comment (roseundy) [reply lib/engine/game/g_1871/meta.rb:19]: OK, thanks.

File: assets/app/view/game/abilities.rb [hunk 2]
```diff
@@ -94,6 +95,20 @@ def actions_for(company)
         @game.round.actions_for(company)
       end
 
+      # Render a button for the purchase train action that purchases the
+      # currently available depot train
+      def render_purchase_train_button
```
Comment (tobymao) [inline assets/app/view/game/abilities.rb:100]: is this needed? why not just use a choice?

File: assets/app/view/game/button/buy_share.rb [hunk 3]
```diff
@@ -16,6 +16,8 @@ class BuyShare < Snabberb::Component
         needs :prefix, default: 'Buy'
         needs :source, default: 'Market'
         needs :action, default: Engine::Action::BuyShares
+        needs :purchase_for, default: nil
```
Comment (tobymao) [inline assets/app/view/game/button/buy_share.rb:19]: why is this needed? is there another way?

File: assets/app/view/game/buy_sell_shares.rb [hunk 4]
```diff
@@ -33,6 +33,8 @@ def render
                            .map(&:first).sort_by(&:percent).reverse
                        end
 
+        @reserved_shares = @corporation.reserved_shares
```
Comment (tobymao) [inline assets/app/view/game/buy_sell_shares.rb:36]: what is this

File: assets/app/view/game/buy_sell_shares.rb [hunk 5]
```diff
@@ -126,6 +131,45 @@ def render_corporate_shares
...
+          @pool_shares.map do |share|
+            next unless @step.can_buy?(target, share.to_bundle, borrow_from: @current_entity)
+
+            h(Button::BuyShare,
+              share: share,
+              entity: @current_entity,
+              purchase_for: target,
+              borrow_from: @current_entity,
+              percentages_available: @pool_shares.size)
+          end + @treasury_shares.map do |share|
```
Comment (tobymao) [inline assets/app/view/game/buy_sell_shares.rb:149]: i don't like this inline style, can you save it to a var and combine?

File: assets/app/view/game/choose.rb [hunk 6]
```diff
@@ -39,6 +48,44 @@ def render
           *choice_buttons,
         ])
       end
+
+      def render_choice_amount(amounts)
+        min = amounts[0]
```
Comment (tobymao) [inline assets/app/view/game/choose.rb:53]: min, max = amounts

File: assets/app/view/game/player.rb [hunk 7]
```diff
@@ -32,6 +33,7 @@ def render
         ]
 
         divs << h(Companies, owner: @player, game: @game, show_hidden: @show_hidden) if @player.companies.any? || @show_hidden
+        divs << h(UnsoldCompanies, owner: @player, game: @game) if @player.unsold_companies.any?
```
Comment (tobymao) [inline assets/app/view/game/player.rb:36]: unless empty?

why is unsold companies needed?

File: assets/app/view/game/player.rb [hunk 8]
```diff
@@ -124,12 +126,17 @@ def render_info
             h(:td, 'Value'),
             h('td.right', @game.format_currency(@game.player_value(@player))),
           ]),
-          h(:tr, [
-            h(:td, 'Liquidity'),
-            h('td.right', @game.format_currency(@game.liquidity(@player))),
-          ]),
         ])
 
+        if @game.show_liquidity?(@player)
```
Comment (tobymao) [inline assets/app/view/game/player.rb:131]: why is this added?

File: assets/app/view/game/split.rb [hunk 9]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+require 'view/game/actionable'
+
+module View
+  module Game
+    class Split < Snabberb::Component
```
Comment (tobymao) [inline assets/app/view/game/split.rb:7]: what does this look like? can you just model this with a choose action?

File: assets/app/view/game/tranches.rb [hunk 10]
```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+require 'lib/settings'
+
+module View
```
Comment (tobymao) [inline assets/app/view/game/tranches.rb:5]: what does this look like and what is it for

File: assets/app/view/game/unsold_companies.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+require 'view/game/company'
+
+module View
+  module Game
+    class UnsoldCompanies < Snabberb::Component
```
Comment (tobymao) [inline assets/app/view/game/unsold_companies.rb:7]: same question

File: lib/engine/ability/purchase_train.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
+  module Ability
+    class PurchaseTrain < Base
```
Comment (tobymao) [inline lib/engine/ability/purchase_train.rb:7]: why is this needed

File: lib/engine/action/buy_shares.rb [hunk 13]
```diff
@@ -5,13 +5,15 @@
 module Engine
   module Action
     class BuyShares < Base
-      attr_reader :entity, :bundle, :swap
+      attr_reader :entity, :bundle, :swap, :purchase_for, :borrow_from
```
Comment (tobymao) [inline lib/engine/action/buy_shares.rb:8]: can you explain this mechanism?

File: lib/engine/action/offer.rb [hunk 14]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
```
Comment (tobymao) [inline lib/engine/action/offer.rb:5]: delete this? we don't need an offer, you can just buy the company

File: assets/app/view/game/abilities.rb (hunk 2 above)
Comment (kelsin) [reply assets/app/view/game/abilities.rb:100]: This is a new company action that lets the owning corporation buy the top train of the depot for normal price at any time. Added here: https://github.com/kelsin/18xx/blob/g_1871/lib/engine/ability/README.md#purchase_train

File: lib/engine/action/purchase_train.rb [hunk 15]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module Engine
```
Comment (tobymao) [inline lib/engine/action/purchase_train.rb:3]: what is this

File: lib/engine/action/offer.rb (hunk 14 above)
Comment (kelsin) [reply lib/engine/action/offer.rb:5]: What do you mean? This is one of the actions in the auction for this game.

File: lib/engine/action/split.rb [hunk 16]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+require_relative 'base'
+
+module Engine
+  module Action
+    class Split < Base
+      attr_reader :corporation
```
Comment (tobymao) [inline lib/engine/action/split.rb:8]: why is this needed? just use convert?

File: lib/engine/action/purchase_train.rb (hunk 15 above)
Comment (kelsin) [reply lib/engine/action/purchase_train.rb:3]: A new ability on companies and the action that goes with it.

https://github.com/kelsin/18xx/blob/g_1871/lib/engine/ability/README.md#purchase_train

File: lib/engine/action/split.rb (hunk 16 above)
Comment (kelsin) [reply lib/engine/action/split.rb:8]: Use the convert action to be the action that happens when you choose to split a corp in 1871?

File: lib/engine/game/base.rb [hunk 17]
```diff
@@ -546,8 +552,12 @@ def inspect
         "#{self.class.name} - #{self.class.title} #{players.map(&:name)}"
       end
 
-      def result
+      def result_players
```
Comment (tobymao) [inline lib/engine/game/base.rb:555]: why is this changed

File: lib/engine/action/buy_shares.rb (hunk 13 above)
Comment (kelsin) [reply lib/engine/action/buy_shares.rb:8]: Ability to buy a share for another player and borrowing money from another player.

The `purchase_for` means it's player A's action, but they are buying a share for player B if you set `purchase_for` to player B. It's used for the union bank in 1871.

The `borrow_from` means if the player that's buying doesn't have enough money you can borrow cash from another player. Also used for the union bank in 1871.

File: lib/engine/game/base.rb [hunk 18]
```diff
@@ -1495,6 +1507,11 @@ def can_par?(corporation, parrer)
         !corporation.ipoed
       end
 
+      # Base definition to not allow splitting in most games
+      def can_split?(_corporation, _parrer)
```
Comment (tobymao) [inline lib/engine/game/base.rb:1511]: why is this needed? if we switch to convert can we get of all of the split code?

File: lib/engine/game/g_1871/game.rb [hunk 19]
```diff
@@ -0,0 +1,822 @@
...
+
+        # We use the reserved share type to represent the 3 shortline exchange
+        # shares.
+        def ipo_reserved_name(_entity = nil)
+          'Exchange'
+        end
+
+        def setup_preround
+          # Pick a random company to be the mainline and the shortline
+          @corporations = @corporations[0..6].sort_by { self.rand } + @corporations[7..-1]
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:98]: def need self

File: lib/engine/game/base.rb (hunk 17 above)
Comment (kelsin) [reply lib/engine/game/base.rb:555]: It's a new method, just git showing the diff weird. the original `result` method is right below.

This is the ability to choose while players are shown in the end game scoring. Union bank is represented here as a player but it shouldn't have it's own score at the end.

File: lib/engine/game/g_1871/game.rb [hunk 20]
```diff
@@ -0,0 +1,822 @@
...
+
+          # Setup the ML and SL names, float values and pars
+          mainline.full_name = mainline.full_name[0..-2] + 'ML'
+          mainline.float_percent = 50
+          shortline.full_name = shortline.full_name[0..-2] + 'SL'
+          stock_market.set_par(mainline, stock_market.share_price(1, 1))
+          stock_market.set_par(shortline, stock_market.share_price(2, 1))
+
+          # Mainline and Shortline tokens start on the board
+          @hexes.find { |hex| hex.id == mainline.coordinates }.tile.cities.first.exchange_token(mainline.tokens.first)
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:120]: just do one for loop and do this

@hexes.each |hex| do 
  if mainline or shortline...

File: lib/engine/game/base.rb (hunk 18 above)
Comment (kelsin) [reply lib/engine/game/base.rb:1511]: We can't get rid of all of the 1871 split code, but yes I should be able to reuse the other action and not have split code in the base.

File: lib/engine/game/g_1871/game.rb [hunk 21]
```diff
@@ -0,0 +1,822 @@
...
+        # Returns the player that should be the current peir owner
+        def peir_owner
+          # Get all of the share holders
+          share_holders = peir.player_share_holders
+
+          # Get max ownership
+          max_ownership = share_holders.map(&:last).max
+
+          # Get the owners that own the max (tied players)
+          max_owners = share_holders.filter { |_p, o| o == max_ownership }.keys
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:347]: never use filter, always use reject

File: lib/engine/game/g_1871/game.rb [hunk 22]
```diff
@@ -0,0 +1,822 @@
...
+
+          # Find the one with the lowest pier share by name
+          max_owners.map { |p| lowest_peir_share(p) }.min_by(&:index).owner
+        end
+
+        # After default caches are made, override the share_by_id and
+        # corporation_by_id method to recognize mainline and shortline
+        def cache_objects
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:355]: is this the only way?

File: lib/engine/game/g_1871/game.rb [hunk 23]
```diff
@@ -0,0 +1,822 @@
...
+        # potential_tiles method which includes 9 as normal but excludes it
+        # based on our upgrades_to? method below.
+        def all_potential_upgrades(tile, tile_manifest: false, selected_company: nil)
+          super.reject { |t| t.name == '9' }
+        end
+
+        # Returns true if this tile represents a X or T hex on the base map
+        def map_x_or_t?(tile)
+          return false unless tile.color == :white
+          return false unless tile.label.to_s.include?('X') || tile.label.to_s.include?('T')
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:417]: don't use unless with multiple conditions

File: lib/engine/game/g_1871/game.rb [hunk 24]
```diff
@@ -0,0 +1,822 @@
...
+          elsif yellow_t?(from)
+            # if we're upgrading a yellow T, make sure we use T tiles
+            return false unless to.label.to_s.include?('T')
+          elsif brown_c?(from)
+            # if we're upgrading a brown C, make sure we use the C tile
+            return false unless to.label.to_s.include?('C')
+          elsif map_x_or_t?(from)
+            # if we're upgrading a map X or T, don't check labels, normal
+            # yellows allowed
+          else
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:469]: elif !map_x_or_t

and delete the else?

File: lib/engine/game/g_1871/game.rb [hunk 25]
```diff
@@ -0,0 +1,822 @@
...
+        # during the stock round.
+        def reorder_players(_order = nil, log_player_order: false)
+          @players.delete(union_bank)
+
+          case @round
+          when init_round.class
+            current_order = @players.dup
+            @players.sort_by! { |p| [p.cash, current_order.index(p)] }
+          else
+            @players = @round.pass_order if @round.pass_order.any?
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:587]: unless empty

File: lib/engine/game/g_1871/game.rb [hunk 26]
```diff
@@ -0,0 +1,822 @@
...
+          return false unless corporation.ipoed
+
+          # Have to own the company in question
+          return false unless corporation.owner == spliter
+
+          # Have to own at least 40 percent
+          return false unless spliter.percent_of(corporation) >= 40
+
+          # Have to have two tokens of this corporation out on the map
+          corporation.placed_tokens.length >= 2
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:634]: s/length/size fix all occurenences

File: lib/engine/game/g_1871/game.rb [hunk 27]
```diff
@@ -0,0 +1,822 @@
...
+        end
+
+        def available_splits
+          corporations.filter do |c|
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:638]: reject

File: lib/engine/game/g_1871/game.rb [hunk 28]
```diff
@@ -0,0 +1,822 @@
...
+            !c.ipoed && c.full_name.include?('Branch')
+          end
+        end
+
+        # Set the PEIR share price and owner based on current shares. Delete the
+        # company if no shares remain
+        def readjust_peir
+          if @peir_shares.length.zero?
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:646]: size

File: lib/engine/game/g_1871/game.rb [hunk 29]
```diff
@@ -0,0 +1,822 @@
...
+            @log << "Closing the King's Mail"
+            company_by_id('KM')&.close!
+            @log << 'Closing the PEIR'
+            peir.close!
+            return
+          end
+
+          new_share_percent = (100 / @peir_shares.length).ceil
```
Comment (tobymao) [inline lib/engine/game/g_1871/game.rb:654]: size

File: lib/engine/game/g_1871/meta.rb [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+
+        GAME_INFO_URL = 'https://github.com/tobymao/18xx/wiki/TheOldPrince'.freeze
+        GAME_LOCATION = 'Prince Edward Island'.freeze
+        GAME_PUBLISHER = nil
```
Comment (tobymao) [inline lib/engine/game/g_1871/meta.rb:23]: we have a self published tag

File: lib/engine/game/g_1871/step/auction.rb [hunk 31]
```diff
@@ -0,0 +1,252 @@
...
+            else
+              # Show the current players pile
+              current_entity.unsold_companies
+            end
+          end
+
+          def active_auction
+            company = @auctioning
+            bids = @bids[company]
+            yield company, bids if bids.any?
```
Comment (tobymao) [inline lib/engine/game/g_1871/step/auction.rb:58]: empty

File: lib/engine/game/g_1871/step/auction.rb [hunk 32]
```diff
@@ -0,0 +1,252 @@
...
+            goto_entity!(@player1)
+          end
+
+          def process_pass(action)
+            player = action.entity
+            @log << "#{player.name} passes on #{@auctioning.name}"
+            player.pass!
+
+            # Are there bids on the item?
+            if @bids[@auctioning].any?
```
Comment (tobymao) [inline lib/engine/game/g_1871/step/auction.rb:97]: empty

File: lib/engine/game/g_1871/step/auction.rb [hunk 33]
```diff
@@ -0,0 +1,252 @@
...
+          end
+
+          def committed_cash(_player, _show_hidden)
+            0
+          end
+
+          def min_bid(company)
+            return unless company
+
+            return starting_bid(company) unless @bids[company].any?
```
Comment (tobymao) [inline lib/engine/game/g_1871/step/auction.rb:144]: empty

File: lib/engine/game/g_1871/step/buy_sell_par_split_shares.rb [hunk 34]
```diff
@@ -0,0 +1,174 @@
...
+
+            entity.companies.each do |company|
+              return true if ["MC", "VR", "SB"].include?(company.id)
+
+              if company.id == "IB"
+                # Only true if there are shares available from an ipo'd company
+                # that isn't PEIR, Mainline or Shortline
+                @game.corporations.select(&:ipoed).reject do |c|
+                  @game.mainline == c or @game.shortline == c or @game.peir == c
+                end.each do |c|
```
Comment (tobymao) [inline lib/engine/game/g_1871/step/buy_sell_par_split_shares.rb:50]: don't like this changed syntax with do end blocks

File: lib/engine/game/g_1871/step/buy_train.rb [hunk 35]
```diff
@@ -0,0 +1,48 @@
...
+          end
+
+          def process_buy_train(action)
+            from_depot = action.train.from_depot?
+            super
+            return unless from_depot
+
+            entity = action.entity
+            @round.bought_trains << entity
+            pass! unless buyable_trains(entity).any?
```
Comment (tobymao) [inline lib/engine/game/g_1871/step/buy_train.rb:25]: empty

File: assets/app/view/game/buy_sell_shares.rb (hunk 4 above)
Comment (kelsin) [reply assets/app/view/game/buy_sell_shares.rb:36]: This game has shares that start reserved, so they aren't buyable. This is part of the code to display those shares (like market / treasury / etc)

File: assets/app/view/game/player.rb (hunk 7 above)
Comment (kelsin) [reply assets/app/view/game/player.rb:36]: During the initial auction you have a pile of companies that you auction off. I wanted to split the display of companies that you are auctioning versus the ones you have bought.

File: assets/app/view/game/player.rb (hunk 8 above)
Comment (kelsin) [reply assets/app/view/game/player.rb:131]: The ability to hide the display of liquidity for certain players. For this game union bank doesn't need that shown. It's a value that's not actionable since the bank can never sell shares.

File: assets/app/view/game/tranches.rb (hunk 10 above)
Comment (kelsin) [reply assets/app/view/game/tranches.rb:5]: This game (like 1835) has tranches, only certain number of companies can open until all companies in the current tranch are either sold out or operated. This displays the tranch chart. We added it to the spreadsheet and stock round views. I can screenshot for you later.

File: assets/app/view/game/split.rb (hunk 9 above)
Comment (kelsin) [reply assets/app/view/game/split.rb:7]: This is the button for split. It's a action you can do on a corp during the stock round.

File: lib/engine/ability/purchase_train.rb (hunk 12 above)
Comment (kelsin) [reply lib/engine/ability/purchase_train.rb:7]: It's a new company ability. The ability to (at any point during the round, but that's handled via the normal action timing) purchase the current train. I added a readme item here: https://github.com/kelsin/18xx/blob/g_1871/lib/engine/ability/README.md#purchase_train

File: lib/engine/game/g_1871/step/buy_sell_par_split_shares.rb (hunk 34 above)
Comment (kelsin) [reply lib/engine/game/g_1871/step/buy_sell_par_split_shares.rb]: Rubocop didn't either, Fixed

File: assets/app/view/game/unsold_companies.rb (hunk 11 above)
Comment (kelsin) [reply assets/app/view/game/unsold_companies.rb:7]: The opening auction has players given a pile of companies that they auction off. This allows people to "own" companies that are waiting to be sold. Probably not applicable many other games, but not able to be represented on screen with any current view options.

File: lib/engine/game/g_1871/game.rb (hunk 22 above)
Comment (kelsin) [reply lib/engine/game/g_1871/game.rb:355]: No, tomorrow I'll experiment with editing abilities to handle this. The idea was that abilities let you exchange for a share of a random company that's set at the start. Instead of editing the abilities during setup I did this, but I can just edit the abilities instead I think, I do that for other reasons anyway.

File: assets/app/view/game/button/buy_share.rb (hunk 3 above)
Comment (kelsin) [reply assets/app/view/game/button/buy_share.rb:19]: This is for adding a button to buy a share for another player. Using it to implement the Union Bank, not another way that I know of since it's a player action that needs to be counted as their buy, but they purchase a share for another player.

File: assets/app/view/game/tranches.rb (hunk 10 above)
Comment (kelsin) [reply assets/app/view/game/tranches.rb:5]: tranches

File: assets/app/view/game/buy_sell_shares.rb (hunk 4 above)
Comment (roseundy) [reply assets/app/view/game/buy_sell_shares.rb:36]: We already have reserved shares. How is this different?

File: assets/app/view/game/tranches.rb (hunk 10 above)
Comment (roseundy) [reply assets/app/view/game/tranches.rb:5]: I used the status field to show this for 1860, 1825, and 18Carolinas.

File: lib/engine/share_pool.rb [hunk 36]
```diff
@@ -29,7 +29,8 @@ def owner
       nil
     end
 
-    def buy_shares(entity, shares, exchange: nil, exchange_price: nil, swap: nil, allow_president_change: true, silent: nil)
+    def buy_shares(entity, shares, exchange: nil, exchange_price: nil, swap: nil,
```
Comment (ventusignis) [inline lib/engine/share_pool.rb:32]: What does the `borrow_from?` parameter mean here?

File: lib/engine/game/g_1871/market.rb [hunk 37]
```diff
@@ -0,0 +1,31 @@
...
+          %w[88 92 100 110 121 133 146 160 180 200 225 250 280],
+          %w[82 86 93 101 111 123 137 152 167 185 203],
+          %w[78 80p 86 94 102 112 122],
+          %w[71 74x 79 84 89 94],
+          %w[60 65z 69 73 78],
+          %w[55 58w 61 65],
+        ].freeze
+
+        # We use par, par_1, par_2, par_3 for our four par options. Use orange
+        # as the 400+40 end square just cause...
```
Comment (ventusignis) [inline lib/engine/game/g_1871/market.rb:18]: Does the "end game blue" (see 1849 for example) color conflict with the par value colors below?

File: lib/engine/game/base.rb (hunk 18 above)
Comment (ventusignis) [reply lib/engine/game/base.rb:1511]: I think you could alternatively remove this and use `if @game.respond_to? :can_split? && @game.can_split?(...)` when you use this method?

File: lib/engine/share_pool.rb (hunk 36 above)
Comment (kelsin) [reply lib/engine/share_pool.rb:32]: It's a new feature added by this PR to borrow money from another player when purchasing this share.

So the action is Player A buys a share with borrow_from: Player B. This means:
- During the check for proper cash, the cache of both players is combined
- When purchasing if Player A doesn't have enough, money is spend from Player B to make up the difference

1871 uses this mechanic for Union Bank purchasing.

File: lib/engine/game/g_1871/market.rb (hunk 37 above)
Comment (kelsin) [reply lib/engine/game/g_1871/market.rb:18]: Sorry I don't understand the question. As far as I know the colors I set below are only used for rendering.

1849 denotes the blue part of their market with `u` which is used for phase limiting: https://github.com/tobymao/18xx/blob/1233d5f60b6973554cc2331de1ec2282c19f0304/lib/engine/share_price.rb#L26

I use the `p,x,z,w` options which are `par,par_1,par_2,par_3`. I didn't have to edit base game code for that. Only set their description strings below and remove those pars on phase events (also doable without editing base game code).

File: lib/engine/game/base.rb (hunk 18 above)
Comment (kelsin) [reply lib/engine/game/base.rb:1511]: Sure, I will edit to that tonight.

File: assets/app/view/game/buy_sell_shares.rb (hunk 4 above)
Comment (kelsin) [reply assets/app/view/game/buy_sell_shares.rb:36]: This is just view code. I needed to display the action to use the private ability to trade for a reserved share, so this is needed to get the reserved shares down when rendering those buttons. I just wanted to follow the pattern of grouping all shares that you needed to render buttons for up here and then displaying them in the render methods below.

File: assets/app/view/game/tranches.rb (hunk 10 above)
Comment (kelsin) [reply assets/app/view/game/tranches.rb:5]: Designer would rather a tranch chart be viewable that try to put this information on each company. Makes it very clear what companies are taking up what tranches and if their are open options. It's only show during the stock rounds and spreadsheet view if the game is using tranches. Right now in 1871 we clean up the entities list during stock rounds to not include all of the companies that can't act / can't open. I'll gladly argue that adding a new module for this is worth it for this game and doesn't muddy up any shared game code.

File: lib/engine/game/g_1871/game.rb (hunk 22 above)
Comment (kelsin) [reply lib/engine/game/g_1871/game.rb:360]: I can change this later if you want but it will be a big change, would rather get this into pre-alpha, get a few spec games going and then be able to safely make the change and know that I didn't break anything.

File: assets/app/view/game/player.rb (hunk 8 above)
Comment (tobymao) [reply assets/app/view/game/player.rb:131]: seems like there's no need to add this then since there's no harm in showing liquidity for union bank, let's delete it?

File: lib/engine/action/offer.rb (hunk 14 above)
Comment (tobymao) [reply lib/engine/action/offer.rb:5]: you don't need to offer something to someone, someone can just buy it instead?

File: assets/app/view/game/tranches.rb (hunk 10 above)
Comment (tobymao) [reply assets/app/view/game/tranches.rb:5]: let's just go with status fields and perhaps the custom html chart that steve added for 21moon?

File: assets/app/view/game/unsold_companies.rb (hunk 11 above)
Comment (tobymao) [reply assets/app/view/game/unsold_companies.rb:7]: does it need to be represented? can you just have an italic or something?

File: lib/engine/game/g_1871/game.rb (hunk 21 above)
Comment (tobymao) [reply lib/engine/game/g_1871/game.rb:352]: looks like you haven't fixed this yet?

File: lib/engine/game/g_1871/game.rb (hunk 22 above)
Comment (tobymao) [reply lib/engine/game/g_1871/game.rb:360]: don't you already have tests cases to add specs for?

Comment (kelsin) [reply lib/engine/game/g_1871/game.rb:360]: Not yet. I have to do auto passing for the auction next, (no code outside of 1871 for that) and it will break any specs I saved for sure, waiting on that.

File: lib/engine/action/offer.rb (hunk 14 above)
Comment (kelsin) [reply lib/engine/action/offer.rb:5]: This auction has all players having a pile of companies that they are have to offer and they choose which item to auction off and then others bid. I just chose "offer" as the action, but if there is another action I can use I can swap to it. It was the first feature I implemented.

File: assets/app/view/game/abilities.rb [hunk 38]
```diff
@@ -94,6 +95,20 @@ def actions_for(company)
...
+        purchase = lambda do
+          process_action(Engine::Action::PurchaseTrain.new(
+            @selected_company
```
Comment (crericha) [inline assets/app/view/game/abilities.rb:103]: Does it make sense to pass the train in as a parameter too, so that the action is more flexible for reuse?

Comment (kelsin) [reply assets/app/view/game/abilities.rb:103]: Sure, but can be added layer once needed.

File: lib/engine/game/base.rb [hunk 39]
```diff
@@ -2055,6 +2074,11 @@ def init_hexes(companies, corporations)
               blockers[hex] = company
             end
           end
+          abilities(company, :blocks_hexes_consent) do |ability|
```
Comment (crericha) [inline lib/engine/game/base.rb:2077]: Why is this a separate ability rather than an option for the block_hexes ability?

File: lib/engine/step/buy_sell_par_shares.rb [hunk 40]
```diff
@@ -139,6 +140,17 @@ def can_sell_order?
...
 
+      def sold_this_turn(corporation)
+        sell_actions = @round.current_actions.select do |a|
+          a.is_a?(Action::SellShares) and a.bundle.corporation == corporation
+        end
+
+        # Find out how much was sold this turn
+        sell_actions.reduce(0) do |total, a|
+          total + a.bundle.percent
+        end
```
Comment (crericha) [inline lib/engine/step/buy_sell_par_shares.rb:151]: I found `sum { |a| a.bundle.percent }` to be easier to read

File: lib/engine/game/base.rb (hunk 39 above)
Comment (kelsin) [reply lib/engine/game/base.rb:2077]: This is only a consent ability, so it's only implemented as the front end check, just like when purchasing trains for others (no real check, just a "Did X say it's ok?". It was less code to make a new ability than to edit the other one and add code to NOT do most of what the other one does in the right places. We could implement it that way for sure, this was simpler.

**Comment by kelsin** [comment]
`show_liquidity` removed and last filter fixed. At this point I'm ready to argue for unsold companies and tranch charts. Let me know what's next.

**Comment by scottredracecar** [comment]
I like the tranche chart and I hope it can be extended to other games like
1860 and 18Carolinas.

**Review by tobymao — APPROVED** [review]
